src/System.Data.SqlClient/src/Microsoft/SqlServer/Server/MetadataUtilsSmi.cs

```diff
@@ -0,0 +1,625 @@
+        //      Given that there's a known SqlDbType, thus a minimal number of comparisions, it's likely this is faster
+        //      than the other approaches considered (both GetType().GetTypeCode() switch and hash table using Type keys
+        //      must instantiate a Type object.  The typecode switch also degenerates into a large if-then-else for
+        //      all but the primitive clr types.
+        internal static ExtendedClrTypeCode DetermineExtendedTypeCodeForUseWithSqlDbType(
+                SqlDbType dbType,
+                bool isMultiValued,
+                object value
+        )
+        {
+            ExtendedClrTypeCode extendedCode = ExtendedClrTypeCode.Invalid;
+
+            // fast-track null, which is valid for all types
+            if (null == value)
+            {
+                extendedCode = ExtendedClrTypeCode.Empty;
+            }
+            else if (DBNull.Value == value)
+            {
+                extendedCode = ExtendedClrTypeCode.DBNull;
+            }
+            else
+            {
+                switch (dbType)
+                {
+                    case SqlDbType.BigInt:
+                        if (value.GetType() == typeof(Int64))
+                            extendedCode = ExtendedClrTypeCode.Int64;
+                        else if (value.GetType() == typeof(SqlInt64))
+                            extendedCode = ExtendedClrTypeCode.SqlInt64;
+                        break;
+                    case SqlDbType.Binary:
+                    case SqlDbType.VarBinary:
+                    case SqlDbType.Image:
+                    case SqlDbType.Timestamp:
+                        if (value.GetType() == typeof(byte[]))
+                            extendedCode = ExtendedClrTypeCode.ByteArray;
+                        else if (value.GetType() == typeof(SqlBinary))
+                            extendedCode = ExtendedClrTypeCode.SqlBinary;
+                        else if (value.GetType() == typeof(SqlBytes))
+                            extendedCode = ExtendedClrTypeCode.SqlBytes;
+                        else if (value.GetType() == typeof(StreamDataFeed))
+                            extendedCode = ExtendedClrTypeCode.Stream;
+                        break;
+                    case SqlDbType.Bit:
+                        if (value.GetType() == typeof(bool))
+                            extendedCode = ExtendedClrTypeCode.Boolean;
+                        else if (value.GetType() == typeof(SqlBoolean))
+                            extendedCode = ExtendedClrTypeCode.SqlBoolean;
+                        break;
+                    case SqlDbType.Char:
+                    case SqlDbType.NChar:
+                    case SqlDbType.NText:
+                    case SqlDbType.NVarChar:
+                    case SqlDbType.Text:
+                    case SqlDbType.VarChar:
+                        if (value.GetType() == typeof(string))
```

Comment:
For these large if/else's, consider calling value.GetType() once and storing the result into a local.